addons/website/static/src/snippets/s_chart/000.js

```diff
@@ -79,18 +80,43 @@ const ChartWidget = publicWidget.Widget.extend({
                 },
             };
         } else {
+            // for extending the range of the axis while maintaining the auto fit behaviour
+            const beforeBuildTicks = (scale) => {
+                scale.min = parseInt(this.el.dataset.minValue) || 0;
+                let maxValue = parseInt(this.el.dataset.maxValue);
+                if (!isNaN(maxValue)) {
+                    scale.max = maxValue;
+                    // for reversing a min max value when min value is greater than max value
+                    if (scale.max < scale.min) {
+                        scale.max = scale.min;
+                        scale.min = maxValue;
+                    } else if (scale.max == scale.min) {
+                        // to maintaining a chart if min value and max value are same for positive and negative number
+                        scale.min < 0 ? (scale.max = 0, scale.min = 2*scale.min) : (scale.min = 0, scale.max = 2*scale.max);
+                    }
+                } else {
+                    // for managing range of axis when max value does not given and min value is greater than chart data value
+                    if (scale.min > Math.max(...data.datasets[0].data)) {
+                        scale.max = scale.min;
+                        scale.min = 0;
+                    }
+                }
+                return;
+            };
```

Comment:
I don't understand why there is so much code for such a small feature... can you explain it to me? Reading the doc https://www.chartjs.org/docs/latest/axes/cartesian/linear.html it seems just about adding a min/max to the "ticks" configuration below... but I probably miss something?

Reply:
Actually as discussed with PO, we want to implement same feature as in google chart for min max. this is code for a different cases of min and max value that's explain in commit message.

Reply:
@qsm-odoo yes, you are right for min & max value.

Reply:
@dwa-odoo @jpr-odoo 
Ok:
- Still, I don't see why you need the `beforeBuildTicks` function. You can compute min/max before, using the new rules and giving it to the lib as with ac767e466bb65ffaa9d91ea48f9f5ab53c25908a I think? But ok just a detail.
- Not a detail: In fact, none of this code should be in s_chart/000.js: you don't need to compute the min/max on each page load of each visitor -> compute the min/max in the option and save it in the DOM correctly (like it would be saved if the user sets it anyway). Maybe I am wrong but I think this will also yield to less code or at least to more readable code ;) That way the public widget here only checks the min/max values which are provided or let the lib do it as before. And also, if we ever want to change those rules for Odoo 17.0... we won't break user snippets as the min/max values using the old rules would have been saved in the DOM.

Reply:
for beforeBuildTicks: if using the new rules and giving it to the lib as with ac767e4 as you said, in that case break the gape between 'two values' in verticle/horizontally axis. it should be same as like google chart
![Home _ My Website](https://user-images.githubusercontent.com/7599806/96979438-c0b15400-153c-11eb-882a-400fec20f35a.png)
![Untitled spreadsheet - Google Sheets](https://user-images.githubusercontent.com/7599806/96979751-d6267e00-153c-11eb-9590-529ed3d08f8f.png)

Reply:
Ok, then add that in a comment otherwise we can't know it reading the code :wink: 
Still: moving the logic to options.js is important :+1:

Reply:
@qsm-odoo we can't use beforeBuildTicks function in option.js file. beacuse of this function will call when build up the chart so, we fix it by function in 000.js file

Reply:
I understand `beforeBuildTicks` must be in 000.js but most of the logic you put in there can be in options.js (like I explained: to avoid loading and executing code for visitors when it is possisble and to avoid potential future compatibility problems). One example to be clear: `if (scale.max < scale.min) {` -> not possible if your options.js had saved its min and max value correctly by inverting them when needed before modifying the DOM.